### ISSUE TITLE
Same Host for multiple Ingress listening on different ports

### ISSUE THREAD

**Issue by hahahannes** [nginxinc/kubernetes-ingress#3313]
**Is your feature request related to a problem? Please describe.**
The goal is to have multiple `VirtualServers` or `Ingresses` that listen on different ports but belong to one host.

## Virtual Servers
My config for one VirtualServer looks like this:
```yml
apiVersion: k8s.nginx.org/v1
kind: VirtualServer
metadata:
  name: vs1
  namespace: server-ns
spec:
  host: foo.bar
  tls:
    secret: tls-secret
  server-snippets: |
        listen 8000 ssl;
  ingressClassName: nginx-internal
  upstreams:
  - name: api
    service: api
    port: 8000
  routes:
  - path: /
    action:
      pass: api
```
It listens on port 8000 and is defined for host `foo.bar`.

I would further like to setup a second VirtualServer that listens on port 8001 on the same host:
```yml
apiVersion: k8s.nginx.org/v1
kind: VirtualServer
metadata:
  name: vs2
  namespace: server-ns
spec:
  host: foo.bar
  tls:
    secret: tls-secret
  server-snippets: |
        listen 8001 ssl;
  ingressClassName: nginx-internal
  policies:
  - name: client-cert-policy
  upstreams:
  - name: api1
    service: api1
    port: 8000
  routes:
  - path: /
    action:
      pass: api1
```
But this setup is not possible as the host is already taken. I followed the recommendation to use `VirtualServers` and `VirtualServerRoutes` for cross-namespace host confiuration (https://github.com/nginxinc/kubernetes-ingress/tree/v2.4.1/examples/custom-resources/cross-namespace-configuration). But this relies on path based routing as only one VirtualServer gets created.

## Ingress
The same problem applies for Ingress. As it not possible anymore to use the same host across ingresses, I followed the recommendation to use Mergable Ingress Ressources (https://github.com/nginxinc/kubernetes-ingress/tree/v2.4.1/examples/ingress-resources/mergeable-ingress-types)

Master Ingress:
```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ing-master
  namespace: server-ns
  annotations:
    kubernetes.io/ingress.class: "nginx-internal"
    nginx.org/mergeable-ingress-type: "master"
spec:
  tls:
    - hosts:
        - foo.bar
      secretName: tls-secret
```

Minion Ingress #1:
```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ing1
  namespace: server-ns
  annotations:
    kubernetes.io/ingress.class: "nginx-internal"
    nginx.org/listen-ports-ssl: "8000"
    nginx.org/mergeable-ingress-type: "minion"
spec:
  rules:
  - host: foo.bar
    http:
      paths:
      - path: /
        pathType: ImplementationSpecific
        backend:
          service:
            name: api1
            port:
              number: 8000
```

Minion Ingress #2:
```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ing2
  namespace: server-ns
  annotations:
    kubernetes.io/ingress.class: "nginx-internal"
    nginx.org/listen-ports-ssl: "8001"
   nginx.org/mergeable-ingress-type: "minion"
spec:
  rules:
  - host: foo.bar
    http:
      paths:
      - path: /
        pathType: ImplementationSpecific
        backend:
          service:
            name: api2
            port:
              number: 8000
```

With this setup, it is possible to have multiple ingress with different ports for one host.
But it will still fail as the path `/` is already occupied by the first minion ingress.

**Describe the solution you'd like**
It would be nice to be able to use the same hostname for multiple ports.

**Describe alternatives you've considered**
This was possible with older versions like `1.8` where it was possible to setup multiple `Ingresses` that would use the same host and would listen on different ports.

Ingress #1
```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ing1
  annotations:
    kubernetes.io/ingress.class: "nginx-internal"
    nginx.org/listen-ports-ssl: "8000"
spec:
  tls:
    - hosts:
        - foo.bar
      secretName: tls-secret
  rules:
  - host: foo.bar
    http:
      paths:
      - path: /
        pathType: ImplementationSpecific
        backend:
          service:
            name: api1
            port:
              number: 8000
```

Ingress #2
```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ing2
  annotations:
    kubernetes.io/ingress.class: "nginx-internal"
    nginx.org/listen-ports-ssl: "8001"
spec:
  tls:
    - hosts:
        - foo.bar
      secretName: tls-secret
  rules:
  - host: foo.bar
    http:
      paths:
      - path: /
        pathType: ImplementationSpecific
        backend:
          service:
            name: api2
            port:
              number: 8000
```

**Additional context**

**Comment by brianehlert** [comment]
This request is blending Layer 4 and Layer 7 together, so you are really going to have to help me understand the problem behind the request.  If you could help me understand the problem behind the feature request, rather than the implementation it would be helpful to me to understand.

Explaining the current state a bit more...
Listener ports happen at layer 4, and those can be defined using a combination of GlobalConfiguration and the deployment/service definition for the ingress controller.
Hostname and path matching happens at layer 7.  Hostname is matched based on the Ingress or VirtualServer object.  Path is matched at the Ingress, VirtualServer or VirtualServerRoute object.

Today, it is possible to have the ingress controller listen on multiple ports and then apply the Layer 7 matching logic.  But it is not possible to one set of Layer 7 matching logic on one port and different Layer 7 matching logic on a different port.
The different Layer 7 rules can be applied based on other Layer 7 aspects such as a header or client IP or other.  Such as doing a Match for something like traffic splitting, A/B, etc.

My perception of the original discussion question was that you wanted both Ingress and VirtualServer to act on the same hostnames.  And that is not possible.  A single hostname must be restricted to one or the other.

**Comment by brianehlert** [comment]
We did add field / parameter validation for the Ingress object specifically.  We did this a couple releases back to improve the security posture.  I wonder if that would be the change.

Could you please share your configurations from before?

**Comment by hahahannes** [comment]
Do you mean my older configuration from the previous version?

If yes, the (minimal) configuration with `1.8.1` looked like this:

```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ing1
  annotations:
    kubernetes.io/ingress.class: "nginx-internal"
    nginx.org/listen-ports-ssl: "8000"
spec:
  tls:
    - hosts:
        - foo.bar
      secretName: tls-secret
  rules:
  - host: foo.bar
    http:
      paths:
      - path: /
        pathType: ImplementationSpecific
        backend:
          service:
            name: api1
            port:
              number: 8000
```

```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ing2
  annotations:
    kubernetes.io/ingress.class: "nginx-internal"
    nginx.org/listen-ports-ssl: "8001"
spec:
  tls:
    - hosts:
        - foo.bar
      secretName: tls-secret
  rules:
  - host: foo.bar
    http:
      paths:
      - path: /
        pathType: ImplementationSpecific
        backend:
          service:
            name: api2
            port:
              number: 8000
```
with both `Ingress` listen on different ports but on the same host.

Thanks!

**Comment by hahahannes** [comment]
I have the feeling that it might help when a hostname collision is only caused by the same hostname and port. Which is the case when one does not set the `nginx.org/listen-ports` or `nginx.org/listen-ports-ssl` annotations

**Comment by hahahannes** [comment]
Ah sorry, I just saw that this checks the rules within one Ingress.

**Comment by hahahannes** [comment]
Hi @brianehlert :)
Just wanted to quickly ask whether I misunderstood something or if I shall clarify it a bit more?

**Comment by brianehlert** [comment]
The project comes from a history of two common HTTP ports ( 80 & 443 ) being used and traffic evaluated using other means such as header, cookie, JWT body, etc.

I also noticed you mentioned v1.8.1
For this project that version was over 2 years ago, and we did not support `nginx.org/listen-ports-ssl` in minions, and still do not today.
https://github.com/nginxinc/kubernetes-ingress/tree/v1.8.1/examples/mergeable-ingress-types

The project currently doesn't support multiple HTTP listeners the way it supports multiple TCP listeners.  We plan on closing this gap during the upcoming year.
I believe this would speak to your use case.  As it would open up the ability to have multiple HTTP listeners.
